cog.c

```diff
@@ -273,6 +268,26 @@ on_handle_local_options (GApplication *application,
                                               ? WEBKIT_TLS_ERRORS_POLICY_IGNORE
                                               : WEBKIT_TLS_ERRORS_POLICY_FAIL);
 
+    if (s_options.filter_path) {
+        WebKitWebsiteDataManager *data_manager =
+            webkit_web_context_get_website_data_manager(cog_shell_get_web_context(shell));
+        g_autofree char *filters_path =
+            g_build_filename(webkit_website_data_manager_get_base_cache_directory(data_manager), "filters", NULL);
+        g_autoptr(WebKitUserContentFilterStore) store = webkit_user_content_filter_store_new(filters_path);
+
+        g_autoptr(GFile) file = g_file_new_for_commandline_arg(s_options.filter_path);
+        g_clear_pointer(&s_options.filter_path, g_free);
+
+        g_autoptr(GMainLoop) loop = g_main_loop_new(NULL, FALSE);
+        webkit_user_content_filter_store_save_from_file(store,
+                                                        "CogFilter",
+                                                        file,
+                                                        NULL,
+                                                        (GAsyncReadyCallback) on_filter_saved,
+                                                        loop);
+        g_main_loop_run(loop);
```

Comment:
Does this work as intended? The GTask spawned in `webkit_user_content_filter_store_save_from_file` will use the thread-default main context, but the global main context will be spun on the basis of passing `NULL` to `g_main_loop_new`.

Should a separate context be spawned and pushed-popped as the thread default? Otherwise spinning the current thread-default one will also be dispatching any other source otherwise associated with execution on this thread (i.e. something in WebKit or any platform-specific source dispatching e.g. windowing events).

Reply:
Currently this seems to works but it may be as well a case of “works on my computer, dunno elsewhere”. I agree that it may be a good idea to go with your approach to make sure we only get the dispatch for the content filter without risking other events being handled as a side effect.

Thanks for the suggestion!

Reply:
Mmmh, amusingly, creating a new `GMainContext` and doing the push/pop dance gets the process stuck. I suspect that inside WebKit we are doing something that uses the _default_ context (as opposed to the _thread default_ context), and by not spinning a main loop on the _default_ one we never get events dispatched—hence the lockup.

Reply:
I think there might be some issue with WebKit's `RunLoop`/`WorkQueue` and how they deal with the _default_ main context and the _thread default_ main context (which are different things); but I do not have time at the moment to investigate further.

Truth is, code works fine using the _default_ main context, so unless somebody can find some case in which this won't work, I would say we merge the PR as-is.

Opinions?

Reply:
Ping? 🏓